File: converted/source2/source/zombie/iso/ChunkSaveWorker.java
package zombie.iso;

import java.io.IOException;
import java.util.ArrayList;
import java.util.concurrent.ConcurrentLinkedQueue;
import zombie.core.Core;
import zombie.debug.DebugLog;
import zombie.network.GameServer;
import zombie.vehicles.BaseVehicle;
import zombie.vehicles.VehiclesDB2;

public class ChunkSaveWorker {
   public static ChunkSaveWorker instance = new ChunkSaveWorker();
   private final ArrayList<IsoChunk> tempList = new ArrayList<>();
   public ConcurrentLinkedQueue<IsoChunk> toSaveQueue = new ConcurrentLinkedQueue<>();
   public boolean bSaving;

   public void Update(IsoChunk var1) {
      if (!GameServer.bServer) {
         IsoChunk var2 = null;
         Object var3 = null;
         this.bSaving = !this.toSaveQueue.isEmpty();
         if (this.bSaving) {
            if (var1 != null) {
               for (IsoChunk var6 : this.toSaveQueue) {
                  if (var6.wx == var1.wx && var6.wy == var1.wy) {
                     var2 = var6;
                     break;
                  }
               }
            }

            if (var2 == null) {
               var2 = this.toSaveQueue.poll();
            } else {
               this.toSaveQueue.remove(var2);
            }

            if (var2 != null) {
               try {
                  var2.Save(false);
               } catch (Exception var5) {
                  var5.printStackTrace();
               }
            }
         }
      }
   }

   public void SaveNow(ArrayList<IsoChunk> var1) {
      this.tempList.clear();

      for (IsoChunk var2 = this.toSaveQueue.poll(); var2 != null; var2 = this.toSaveQueue.poll()) {
         boolean var3 = false;

         for (int var4 = 0; var4 < var1.size(); var4++) {
            IsoChunk var5 = (IsoChunk)var1.get(var4);
            if (var2.wx == var5.wx && var2.wy == var5.wy) {
               try {
                  var2.Save(false);
               } catch (IOException var7) {
                  var7.printStackTrace();
               }

               var3 = true;
               break;
            }
         }

         if (!var3) {
            this.tempList.add(var2);
         }
      }

      for (int var8 = 0; var8 < this.tempList.size(); var8++) {
         this.toSaveQueue.add(this.tempList.get(var8));
      }

      this.tempList.clear();
   }

   public void SaveNow() {
      DebugLog.log("EXITDEBUG: ChunkSaveWorker.SaveNow 1");

      for (IsoChunk var1 = this.toSaveQueue.poll(); var1 != null; var1 = this.toSaveQueue.poll()) {
         try {
            DebugLog.log("EXITDEBUG: ChunkSaveWorker.SaveNow 2 (ch=" + var1.wx + ", " + var1.wy + ")");
            var1.Save(false);
         } catch (Exception var3) {
            var3.printStackTrace();
         }
      }

      this.bSaving = false;
      DebugLog.log("EXITDEBUG: ChunkSaveWorker.SaveNow 3");
   }

   public void Add(IsoChunk var1) {
      if (Core.getInstance().isNoSave()) {
         for (int var2 = 0; var2 < var1.vehicles.size(); var2++) {
            VehiclesDB2.instance.updateVehicle((BaseVehicle)var1.vehicles.get(var2));
         }
      }

      if (!this.toSaveQueue.contains(var1)) {
         this.toSaveQueue.add(var1);
      }
   }
}

File: ProjectZomboidSwitch/src/zombie/iso/ChunkSaveWorker.cpp

#include "zombie/iso/ChunkSaveWorker.h"
#include "zombie/iso/IsoChunk.h"
#include "zombie/core/Core.h"
#include "zombie/vehicles/BaseVehicle.h"
#include "zombie/vehicles/VehiclesDB2.h"
#include <algorithm>
#include <iostream>

namespace zombie {
namespace iso {

std::shared_ptr<ChunkSaveWorker> ChunkSaveWorker::instance = std::make_shared<ChunkSaveWorker>();

ChunkSaveWorker::ChunkSaveWorker() : bSaving(false) {}
ChunkSaveWorker::~ChunkSaveWorker() {}

void ChunkSaveWorker::Update(std::shared_ptr<IsoChunk> chunk) {
    if (GameServer::bServer) return;
    std::lock_guard<std::mutex> lock(queueMutex);
    std::shared_ptr<IsoChunk> found = nullptr;
    bSaving = !toSaveQueue.empty();
    if (bSaving) {
        if (chunk) {
            for (const auto& c : toSaveQueue) {
                if (c && c->wx == chunk->wx && c->wy == chunk->wy) {
                    found = c;
                    break;
                }
            }
        }
        if (!found && !toSaveQueue.empty()) {
            found = toSaveQueue.front();
            toSaveQueue.pop_front();
        } else if (found) {
            toSaveQueue.erase(std::remove(toSaveQueue.begin(), toSaveQueue.end(), found), toSaveQueue.end());
        }
        if (found) {
            try {
                found->Save(false);
            } catch (const std::exception& e) {
                std::cerr << "ChunkSaveWorker::Update exception: " << e.what() << std::endl;
            }
        }
    }
}

void ChunkSaveWorker::SaveNow(const std::vector<std::shared_ptr<IsoChunk>>& chunks) {
    std::lock_guard<std::mutex> lock(queueMutex);
    tempList.clear();
    while (!toSaveQueue.empty()) {
        auto c = toSaveQueue.front();
        toSaveQueue.pop_front();
        bool matched = false;
        for (const auto& chk : chunks) {
            if (c && chk && c->wx == chk->wx && c->wy == chk->wy) {
                try {
                    c->Save(false);
                } catch (const std::exception& e) {
                    std::cerr << "ChunkSaveWorker::SaveNow exception: " << e.what() << std::endl;
                }
                matched = true;
                break;
            }
        }
        if (!matched) tempList.push_back(c);
    }
    for (const auto& c : tempList) {
        toSaveQueue.push_back(c);
    }
    tempList.clear();
}

void ChunkSaveWorker::SaveNow() {
    DebugLog::log("EXITDEBUG: ChunkSaveWorker.SaveNow 1");
    std::lock_guard<std::mutex> lock(queueMutex);
    while (!toSaveQueue.empty()) {
        auto c = toSaveQueue.front();
        toSaveQueue.pop_front();
        try {
            DebugLog::log("EXITDEBUG: ChunkSaveWorker.SaveNow 2 (ch=" + std::to_string(c->wx) + ", " + std::to_string(c->wy) + ")");
            c->Save(false);
        } catch (const std::exception& e) {
            std::cerr << "ChunkSaveWorker::SaveNow exception: " << e.what() << std::endl;
        }
    }
    bSaving = false;
    DebugLog::log("EXITDEBUG: ChunkSaveWorker.SaveNow 3");
}

void ChunkSaveWorker::Add(std::shared_ptr<IsoChunk> chunk) {
    if (Core::getInstance()->isNoSave()) {
        for (const auto& v : chunk->vehicles) {
            VehiclesDB2::instance->updateVehicle(std::dynamic_pointer_cast<BaseVehicle>(v));
        }
    }
    std::lock_guard<std::mutex> lock(queueMutex);
    auto it = std::find(toSaveQueue.begin(), toSaveQueue.end(), chunk);
    if (it == toSaveQueue.end()) {
        toSaveQueue.push_back(chunk);
    }
}

} // namespace iso
} // namespace zombie
